Mathlib/CategoryTheory/Distributive/Monoidal.lean

```diff
@@ -0,0 +1,230 @@
+
+- The category of abelian groups with the monoidal structure given by the tensor product of
+  abelian groups.
+- The category of R-modules with the monoidal structure given by the tensor product of modules.
+- The category of vector bundles over a topological space where the monoidal structure is given by
+  the tensor product of vector bundles.
+- The category of pointed types with the monoidal structure given by the smash product of
+  pointed types and the coproduct given by the wedge sum.
+
+## References
+
+[Hans-Joachim Baues, Mamuka Jibladze, Andy Tonks, Cohomology of
+ monoids in monoidal categories, in: Operads: Proceedings of Renaissance
+ Conferences, Contemporary Mathematics 202, AMS (1997) 137-166][]
+
+-/
+
+universe v v₂ u u₂
+
+noncomputable section
+
+namespace CategoryTheory
+
+open Category MonoidalCategory Limits Iso
+
+variable {C : Type u} [Category.{v} C] [MonoidalCategory.{v} C] [HasBinaryCoproducts C]
+
+/-- The canonical left distributivity morphism -/
+def distributorLeft (X Y Z : C) : (X ⊗ Y) ⨿ (X ⊗ Z) ⟶ X ⊗ (Y ⨿ Z) :=
+  coprod.desc (_ ◁ coprod.inl) (_ ◁ coprod.inr)
+
+/-- The canonical right distributivity morphism -/
+def distributorRight (X Y Z : C) : (Y ⊗ X) ⨿ (Z ⊗ X) ⟶ (Y ⨿ Z) ⊗ X :=
+  coprod.desc (coprod.inl ▷ _) (coprod.inr ▷ _)
+
+/-- Notation for the left distributor. -/
+notation "∂L " => distributorLeft
+
+/-- Notation for the right distributor. -/
+notation "∂R " => distributorRight
+
+@[reassoc (attr := simp)]
+lemma coprod_inl_comp_distributorLeft {X Y Z : C} : coprod.inl ≫ (∂L X Y Z) = (X ◁ coprod.inl) :=
+by
+  rw [distributorLeft, coprod.inl_desc]
+
+@[reassoc (attr := simp)]
+lemma coprod_inr_comp_distributorLeft {X Y Z : C} : coprod.inr ≫ (∂L X Y Z) = (X ◁ coprod.inr) :=
+by
+  rw [distributorLeft, coprod.inr_desc]
+
+@[reassoc (attr := simp)]
+lemma coprod_inl_comp_distributorRight {X Y Z : C} : coprod.inl ≫ (∂R X Y Z) = (coprod.inl ▷ X) :=
+by
+  rw [distributorRight, coprod.inl_desc]
+
+@[reassoc (attr := simp)]
+lemma coprod_inr_comp_distributorRight {X Y Z : C} : coprod.inr ≫ (∂R X Y Z) = (coprod.inr ▷ X) :=
+by
+  rw [distributorRight, coprod.inr_desc]
+
+/-- In a ymmetric monoidal category, the right distributor equals
+the left distributor up to braidings. -/
+@[reassoc (attr := simp)]
+lemma right_distributor_braiding_left_distributor [SymmetricCategory C]
+    {X Y Z : C} :
+  ∂R X Y Z ≫ (β_ (Y ⨿ Z) X).hom = (coprod.map (β_ Y X).hom (β_ Z X).hom) ≫ ∂L X Y Z := by
+    ext
+    · simp
+    · simp
+
+@[reassoc (attr := simp)]
+lemma left_distributor_braiding_right_distributor [SymmetricCategory C]
+    {X Y Z : C} :
+  ∂L X Y Z ≫ (β_ X (Y ⨿ Z)).hom = (coprod.map (β_ X Y).hom (β_ X Z).hom) ≫ ∂R X Y Z := by
+    ext
+    · simp
+    · simp
+
+variable (C)
+
+/-- A monoidal category with binary coproducts is left distributive
+if the left distributor is an isomorphism. -/
+class MonoidalLeftDistributive where
+  is_iso_distributorLeft {X Y Z : C} : IsIso (∂L X Y Z) := by infer_instance
```

Comment:
I think this would be a better design:
```suggestion
class MonoidalLeftDistributive : Prop where
  preservesBinaryCoproducts_tensorLeft (X : C) : 
    PreservesLimitsOfShape (Discrete WalkingPair) (tensorLeft X)
```

Reply:
Can you elaborate? is it because `PreservesLimitsOfShape` gives better automatic instance inference or some other reason?

Reply:
By the way in few places, e.g. in `Distributive.Cartesian` the instance which proves `MonoCoprod C` from `MonoidalLeftDistributive C`, I explicitly use the distributor isomorphisms `∂L`. Going with your suggestion, one thing we can do is to explicitly prove the invertibility of `∂L` from preservation of coproducts. Would that be a good API?

Reply:
We have a reasonably good API for the preservation of limits and colimits: we should use it, and improve it if this is necessary. One of the good reasons to use this approach is that in the closed monoidal case, we immediately get the preservations of binary coproducts from the fact `tensorLeft` is a left adjoint. I would suggest defining `∂L` as an isomorphism (using `PreservesColimitPair.iso`) under the assumption `MonoidalLeftDistributive`.

Reply:
Great! Just made changes to implement this idea. In particular `∂L` is now defined as an isomorphism form `PreservesColimitPair.iso`. And the proof of of `IsMonoidalLeftDistrib` from `MonoidalClosed` uses `(ihom.adjunction X).leftAdjoint_preservesColimits`.